### ISSUE TITLE
New Page: Login (+registration & forgotten password)

### ISSUE THREAD

**Issue by mrlacey** [Microsoft/WindowsTemplateStudio#291]
Add pages for Login, forgot password & basic registration. for apps that need to authenticate with a private backend (XAuth style) 
This should have placeholder service for performing the actual authentication and documentation on what should be implemented by the developer.
This also needs to support changing the start page of the generated project so that must log in before getting to the actual app pages.
For NavPanel template the login pages should not be in the shell. The secondary menu should add a "log out" option.
X-Ref #130 

Note this requires adding multiple pages at the same time. Replacements from the template config aren't currently applied to all pages yet and so this must be supported too.

**Comment by mrlacey** [comment]
I have a PoC of this and have shown it to Clint & Michael. Will add as a branch next week when I get back from build.

**Comment by crutkas** [comment]
I can't remember who i was talking to but things that it should have would be forgot password and new user as well

**Comment by mrlacey** [comment]
@crutkas by "registration" I meant "new user". Forgotten password already covered

**Comment by mrlacey** [comment]
For reference this needs #294 completing first

**Comment by dotMorten** [comment]
I'd would want this as a landing page - the rest of the app isn't accessible until you log in. Log out brings you back to this page. This seems to match the behavior of almost all apps that requires a signin (twitter, facebook, slack etc etc)

Code-behind implementation could just be something as simple as a fake "login" that just sets a username from what was typed in the username box, with a `// TODO` to provider "proper" authentication.

**Comment by mrlacey** [comment]
@dotMorten yes, the intention is that this would be the first page displayed when the app was launched. 
Launching gets more complicated when combined with an extended splash screen (#293) and that it's now possible to change the "home" page (#294)
The logic for login will be stubbed out with comments but there needs to be a balance between doing too much and too little. I don't want what's provided to be seen as good enough but also don't want to force a specific way of doing it or leaving lots of code for some people to take out.

**Comment by mrlacey** [comment]
Removing myself as assignee as don't want to block others from moving this forward.

**Comment by mrlacey** [comment]
The big blocker for this is **validation**.

You can't create these pages without some form of validation.  
**BUT** there is no validation functionality built-in to the platform and no standards for how this is broadly implemented by the wider development community.

There are two parts to validation.
1. The logic to assess the entered data.
2. Indication of any issues in the UI.

There is an [official uservoice request](https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/13052589-uwp-input-validation) for this but doesn't appear to be something coming soon.

Prism includes some built-in validation functionality. Caliburn.Micro deliberately doesn't and leaves it up to the individual developer. The other framework options we have include nothing specifically for this.

There are lots of third-party solutions of varying popularity.  
For completeness, it's probably worth mentioning [Template10.Validation](https://github.com/Windows-XAML/Template10.Validation) as it is especially UWP focused but this has not been updated in the last 6 months.  
The most popular (by my reckoning of nuget download numbers) general purpose validation framework is [FluentValidation](https://github.com/JeremySkinner/FluentValidation) which may also be worth considering.

**Comment by emiliano84** [comment]
@mrlacey hope to not bother you guys, but maybe you can take a look at my form validation for UWP, I didn't have time to finish to work on it and write down documentation, also probably is not the best code to work with, but maybe can help as idea and/or base for the development. In any case any suggestion will be appreciated :)

https://github.com/emiliano84/FormValidationUwp

**Comment by ralarcon** [comment]
Before 17'Christmas, we were implementing a PoC for Authentication using web authentication broker + custom auth that includes a login and register page (not recovery).

https://github.com/mvegaca/WindowsTemplateStudio/tree/authentication-poc   

Some thoughts regarding this PoC:
The idea behind this PoC is to have an "Authentication Provider" mechanism which allows defining different providers.

All providers extend from AuthenticationProviderBase and must to implement the abstract method:
Task<AuthenticationResult> AuthenticateAsync(); Based on that, the View model can handle different providers that use this protocol and UI can have different buttons for each Login Provider.

Social providers (Twitter, Facebook, and Google) providers are using WebAuthenticationBroker API that shows a pop-up to introduce the social network credentials and returns the authentication result and authentication data (access token, user id, …). In case of Microsoft Account the provider use Windows.UI.ApplicationSettings.AccountsSettingsPane to get Microsoft Account Login.
This protocol is flexible to add different providers using other technologies like Azure AD, GitHub login,…

Other interesting points in App:
-	Include styles in resource dictionaries.
-	Include error handling and error user feedback.

We were thinking about how to translate this PoC to templates in WTS. There could be different approaches to translate this implantation, but here is one approach that could help:

In WTS, the user can choose one or more Login features; all of these features depends on Authentication Basic Component. Login feature will run merge post action in Basic component feature and will also add new files.

Regarding validations, the PoC includes some but we didn't investigate too much around it.

**Comment by mrlacey** [comment]
@ralarcon that POC looks to be a hybrid of functionality defined in #292, #331 and this issue. Big question: why wasn't the existence of the POC documented in any of them? What else has been investigated or had a POC created but there's no documentation of it in an issue?

**Comment by ralarcon** [comment]
That was just an initial approach. You are right it is a hybrid of those issues. I thought we mentioned it on the issues and that was my fault because of I didn't ensure it was mentioned.

**Comment by ralarcon** [comment]
By the way, I updated #297 to mention that we have other PoC too.
